src/vm/arm/asmhelpers.S

```diff
@@ -0,0 +1,1358 @@
+        cmp \ptrReg, __wbScratch
+        blo \label
+        LOAD_GC_GLOBAL \name, __wbScratch, g_highest_address
+        cmp \ptrReg, __wbScratch
+        bhs \label
+    .endm
+    
+    .macro JIT_WRITEBARRIER name, mp, post
+    LEAF_ENTRY \name, _TEXT
+        START_WRITE_BARRIER \name
+        .if(\mp)
+            dmb
+        .endif
+        
+        str r1, [r0]
+        UPDATE_GC_SHADOW \name, r0, r1
+        UPDATE_CARD_TABLE \name, r0, r1, \mp, \post, r0
+        bx lr
+    LEAF_END \name, _TEXT
+    .endm
+    
+    .macro JIT_CHECKEDWRITEBARRIER_SP name, post
+    LEAF_ENTRY \name, _TEXT
+        START_WRITE_BARRIER \name
+        str r1, [r0]
+        CHECK_GC_HEAP_RANGE \name, r0, 1f
+        UPDATE_GC_SHADOW \name, r0, r1
+        UPDATE_CARD_TABLE \name, r0, r1, 0, \post, r0
+1:
+        bx lr
+    LEAF_END \name, _TEXT
+    .endm
+    
+    .macro JIT_CHECKEDWRITEBARRIER_MP name, post
+    LEAF_ENTRY \name, _TEXT
+        START_WRITE_BARRIER \name
+        dmb
+        str r1, [r0]
+        CHECK_GC_HEAP_RANGE \name, r0, 1f
+        UPDATE_GC_SHADOW \name, r0, r1
+        UPDATE_CARD_TABLE \name, r0, r1, 1, \post, r0
+        bx lr
+1:
+        str r1, [r0]
+        bx lr
+    LEAF_END \name, _TEXT
+    .endm
+    
+    .macro JIT_BYREFWRITEBARRIER name, mp, post
+    LEAF_ENTRY \name, _TEXT
+        START_WRITE_BARRIER \name
+        .if(\mp)
+            dmb
+        .endif
+        
+        ldr r2, [r1]
+        str r2, [r0]
+        CHECK_GC_HEAP_RANGE \name, r0, 1f
+        UPDATE_GC_SHADOW \name, r0, r2
+        UPDATE_CARD_TABLE \name, r0, r2, \mp, \post, r2
+1:
+        add r0, #4
+        add r1, #4
+        bx lr
+    LEAF_END \name, _TEXT
+    .endm
+    
+    .macro JIT_WRITEBARRIER_DESCRIPTOR name
+        .word \name
+        .word \name\()_End
+        .word __\name\()__g_lowest_address_offset
+        .word __\name\()__g_highest_address_offset
+        .word __\name\()__g_ephemeral_low_offset
+        .word __\name\()__g_ephemeral_high_offset
+        .word __\name\()__g_card_table_offset
+    .endm
+
+    // There 4 versions of each write barriers. A 2x2 combination of multi-proc/single-proc and pre/post grow version
+    JIT_WRITEBARRIER JIT_WriteBarrier_SP_Pre,  0, 0
+    JIT_WRITEBARRIER JIT_WriteBarrier_SP_Post, 0, 1
+    JIT_WRITEBARRIER JIT_WriteBarrier_MP_Pre,  1, 0
+    JIT_WRITEBARRIER JIT_WriteBarrier_MP_Post, 1, 1
+    
+    JIT_CHECKEDWRITEBARRIER_SP JIT_CheckedWriteBarrier_SP_Pre,  0
+    JIT_CHECKEDWRITEBARRIER_SP JIT_CheckedWriteBarrier_SP_Post, 1
+    JIT_CHECKEDWRITEBARRIER_MP JIT_CheckedWriteBarrier_MP_Pre,  0
+    JIT_CHECKEDWRITEBARRIER_MP JIT_CheckedWriteBarrier_MP_Post, 1
+
+    JIT_BYREFWRITEBARRIER JIT_ByRefWriteBarrier_SP_Pre,  0, 0
+    JIT_BYREFWRITEBARRIER JIT_ByRefWriteBarrier_SP_Post, 0, 1
+    JIT_BYREFWRITEBARRIER JIT_ByRefWriteBarrier_MP_Pre,  1, 0
+    JIT_BYREFWRITEBARRIER JIT_ByRefWriteBarrier_MP_Post, 1, 1
+    
+//    .section .clrwb, "d"
```

Comment:
I didn't debug too far and perhaps it would simply be some linker options, but I was getting garbage when I put it in it's own section. It seems that the section was only required with the previous code anyway because the table was generated in the definitions, and with GNU assembler syntax I couldn't work out an equivalent way of implementing it like that, this was the closest I could get and is functionally equivalent.